### ISSUE TITLE
Add non_blocking=True copy of torch.Storage

### ISSUE THREAD

**Issue by y1r** [pytorch/pytorch#42652]
## 🐛 Bug
torch.Storage.copy_ has `non_blocking` parameter but there is no implementation of async copy.
Therefore, this copy request will be fall-backed to synced copy.

## Motivation of async use
I want to overwrite the activation during backprop for some reason.
Since PyTorch checks the tensor version if performs in-place operation, `tensor.detach().copy_` also doesn't work.

## To Reproduce
```py
a = torch.empty(10, device='cuda')
b = torch.empty(10, device='cuda')

b.storage().copy_(a.storage(), non_blocking=True) # blocks by cudaStreamSynchronize
```

## Expected behavior
No synchronization.

## Environment
- PyTorch Version: 9add11ffc143257fe1e3579702155f39eed85dc4 (HEAD)
- OS: Linux
- Build command you used
```
git clone git://github.com/pytorch/pytorch
USE_MKLDNN=0 USE_NNPACK=0 USE_QNNPACK=0 BUILD_CAFFE2_OPS=0 pip install -e .
```
- CUDA/cuDNN version: cuda-10.2, cudnn 7
- GPU models and configuration: RTX 2060 SUPER

## Patch
I've already implemented async copy under THC. If okay, I'll open PR.
https://github.com/y1r/pytorch/commit/8fded05fe91b0fb4ba527678f84b3f898b295251

**Comment by ailzhang** [comment]
Why does this feature request is on torch.Storage instead of torch.Tensor btw?

**Comment by y1r** [comment]
I want to overwrite the activation during backprop for some reason.
Since `torch.Tensor` tracks the number of in-place updates, modifying them in `torch.Tensor` API increments internal Tensor version so this [error](https://discuss.pytorch.org/t/encounter-the-runtimeerror-one-of-the-variables-needed-for-gradient-computation-has-been-modified-by-an-inplace-operation/836) will occur. However, overwriting torch.Storage is okay because internal tensor version will not be incremented.

**Comment by kshitij12345** [comment]
@y1r 
Just curious. If you actually manage to overwrite some activation in middle layer, wouldn't the gradient that will flow to the layers before be incorrect?

**Comment by ailzhang** [comment]
Yea as @kshitij12345 suggested, we probably need to understand the use case better before proceeding with this feature request. (In general we take `torch.Tensor` as public facing API and `torch.Storage` is kinda an internal concept).

**Comment by y1r** [comment]
@kshitij12345 @ailzhang Thank you for your response.

Currently, I am working on a memory-efficient backprop. In order to reduce the GPU memory consumption, my implementation sometimes moves activation tensor to CPU and move tensor from CPU to GPU if necessary.

Example for description
```py
inp = torch.rand(10, device='cuda', requires_grad=True)
output = torch.relu(inp)

output_cpu = output.detach().cpu()
print(output_cpu)
output.storage().resize_(0)
output.storage().resize_(10)
output.detach().copy_(output_cpu)

print(loss)

loss.backward()
print(inp.grad)
```
This example works, but actually I used hook function to perform these swap-in and swap-out during forward and backward propagation. In some cases, these copy increments internal tensor version so the backprop fails. However, torch.Storage copy doesn't increment the tensor version, so I employed `torch.Storage.copy_`.
